### ISSUE TITLE
Decide on semantics for enum -> integer casts on non-Copy enums.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#35941]
Casting an `enum` which does not derive `Copy` to an integer moves atm:

``` rust
enum E {
    A = 0,
    B = 1,
    C = 2
}

fn main() {
    let e = E::C;
    assert_eq!(e as u32, 2);
    assert_eq!(e as u32, 2); // error: use of moved value: `e`
}
```

This matches the behavior of other coercions/casts, specifically unsizing ones (`Box<T> -> Box<Trait>`).

However, one difference arises between old trans and MIR trans, wrt drops:

``` rust
enum E {
    A = 0,
    B = 1,
    C = 2
}

impl Drop for E {
    fn drop(&mut self) {
        println!("drop");
    }
}

fn main() {
    let e = E::C;
    assert_eq!(e as u32, 2);
}
```

On stable, old trans will run the destructor exactly once. After #35764, MIR trans doesn't run it at all.
The reason is that the move that borrowck knows about causes the drops in MIR to be statically elided, whereas old trans forgot(?) to drop-fill `e`, resulting in the variable being dropped at end of scope.

Should the drop execute or not? IMO it's good to have uniformity between all moving operations.

**Comment by Aatch** [comment]
The safe option is, of course, to execute the drop. However I think that not running makes more sense to me. For one, it would match the semantics of transmute, which is the only other way to get this kind of situation (converting from a type that impls to one that doesn't).

Another option, though it would be more breaking than either already mentioned, would be to disallow enum -> int casts where the enum implements `Drop`.

**Comment by arielb1** [comment]
My original intent when I wrote the conversion code was to execute the drop. That was before `mem::forget` was made safe, so it was the only option.

**Comment by nikomatsakis** [comment]
We discussed this in the @rust-lang/lang meeting. We like the idea of making it an error to coerce an enum to an integer if that enum type supports `Drop` -- presuming it doesn't have impact in practice (which seems likely).

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by oli-obk** [comment]
Some mentoring instructions of how I think this could be implemented:

1. Turn the match arm body https://github.com/rust-lang/rust/blob/3bf71b3d5c07afc0908881a0868080ed399ca6ca/src/librustc_typeck/check/cast.rs#L619 into a function call similar to all the `check_X_Y_cast` functions.
2. Add check whether the `self.expr_ty` has a destructor (you can use the `has_dtor` meethod for this, but that method takes a `TyCtxt` so you need to pass an `fcx` to your new function just like some of the other `check_X_Y_cast` functions do.
3. If the type has a destructor, create a new error variant in https://github.com/rust-lang/rust/blob/3bf71b3d5c07afc0908881a0868080ed399ca6ca/src/librustc_typeck/check/cast.rs#L152 and report that. This will also require you to actually emit an error in https://github.com/rust-lang/rust/blob/3bf71b3d5c07afc0908881a0868080ed399ca6ca/src/librustc_typeck/check/cast.rs#L217 (the compiler will tell you to handle the new error variant). Create a structured error similar to how https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/cast.rs#L358-L366 works.

**Comment by oddg** [comment]
Thanks @oli-obk for the detailed approach. I'll take a stab at it if no one is working on it already.

**Comment by JohnTitor** [comment]
I'm going to close this in favor of #72331 which implemented `cenum_impl_drop_cast` and #73333 which is tracking it. Feel free to re-open if I'm missing something.
